# General description 

Script and associated data used to benchmark sets of methods frequently employed to analyse eDNA sequencing outputs.

# Required files, scripts and programs

## Programs

- seqkit v2.1 ([website](https://bioinf.shenwei.me/seqkit/))
- RDP tool classifier v2.11 ([Github](https://github.com/rdpstaff/classifier) / [SourceForge](https://sourceforge.net/projects/rdp-classifier/))
- usearch v11.0.667 ([website](http://www.drive5.com/usearch/))
- vsearch v2.15.2 ([Github](https://github.com/torognes/vsearch))
- Barque V1.8.5 ([Github](https://github.com/enormandeau/barque)) (N.B. Barque will only work on GNU Linux or OSX). Requires the following dependencies : 
  - bash 4+
  - python 3.5+
  - python distutils package
  - R 3+ (ubuntu/mint: sudo apt-get install r-base-core)
  - java (ubuntu/mint: sudo apt-get install default-jre)
  - gnu parallel
  - flash (read merger) v1.2.11+
  - vsearch v2.14.2+ (Barque will not work with older versions of vsearch)

## Databases

The following databases can be downloaded with clickiing the hyperlinks

- [COI formatted for Barque](https://www.ibis.ulaval.ca/services/bioinformatique/barque_databases/)
- [Eukaryote CO1 Classifier for RDP V5.1](https://github.com/terrimporter/CO1Classifier/releases/tag/RDP-COI-v5.1.0) 
- [12S fish Classifier v1.0.1](https://github.com/terrimporter/12SfishClassifier/releases/tag/v1.0.1)
- [12S formatted for barque](https://github.com/enormandeau/barque/blob/master/03_databases/12S.fasta.gz)

The sequence header of each database (required for certain script) can be extracted using `grep -e ">" databasename.fasta > header_db_markergene.txt` 

The following nomenclature was used for each database. Generated files should then be place in folder `database`

- COI formatted for Barque : `header_bold_coi_barque.txt`
- Eukaryote CO1 Classifier for RDP V5.1 : `header_rdp_coi.txt`
- 12S fish Classifier v1.0.1 : `header_rdp_12S.txt`
- 12S formatted for barque : `header_barque_12S.txt`
 
# Script description 

[00_instructions_Barque.Rmd](https://github.com/karinevilleneuve/benchmarking_eDNA_method_AlbertaFish/blob/main/00_instructions_Barque.Rmd)
- Instructions for executing the Barque pipeline

[00_instructions_insilico_communities.Rmd](https://github.com/karinevilleneuve/benchmarking_eDNA_method_AlbertaFish/blob/main/00_instructions_insilico_communities.Rmd)
- Instructions for generating *in-silico* FASTQ using the ART software and also InsilicoSeq
  
[00_instructions_train_curatedDB_4_RDP.Rmd](https://github.com/karinevilleneuve/benchmarking_eDNA_method_AlbertaFish/blob/main/00_instructions_train_curatedDB_4_RDP.Rmd)
- Instructions for training the curated database in order to use them with RDP

[00_rawDatabase_piechart.R](https://github.com/karinevilleneuve/benchmarking_eDNA_method_AlbertaFish/blob/main/00_rawDatabase_piechart.R)
- Script to assess the representation of taxa considered as fish in the ready-to use databases

[01_DADA2_12S.R](https://github.com/karinevilleneuve/benchmarking_eDNA_method_AlbertaFish/blob/main/01_DADA2_12S.R): Script used to process the FASTQs generated from sequencing the 12S marker gene  
[01_DADA2_COI.R](https://github.com/karinevilleneuve/benchmarking_eDNA_method_AlbertaFish/blob/main/01_DADA2_COI.R): Script used to process the FASTQs generated from sequencing the COI marker gene 
[02_parsing_RDP_output.R](https://github.com/karinevilleneuve/benchmarking_eDNA_method_AlbertaFish/blob/main/02_pasing_RDP_output.R): Script used to parse the output files from the RDP classifier 
[03_combine12S.R](https://github.com/karinevilleneuve/benchmarking_eDNA_method_AlbertaFish/blob/main/03_combine12S.R): Script used to combine results from all methods for the samples sequenced using the 12S marker gene 
[03_combineCOI.R](https://github.com/karinevilleneuve/benchmarking_eDNA_method_AlbertaFish/blob/main/03_combineCOI.R): Script used to combine results from all methods for the samples sequenced using the COI marker gene 
[04_ASV_and_AlbertaSpecies_perSample.R](https://github.com/karinevilleneuve/benchmarking_eDNA_method_AlbertaFish/blob/main/04_ASV_and_AlbertaSpecies_perSample.R): Script used to assess the total number of ASVs generated by each method and the number of distinct Alberta freshwater fish species identified
[04_bootstrapping_cutoffs.R](https://github.com/karinevilleneuve/benchmarking_eDNA_method_AlbertaFish/blob/main/04_bootstrapping_cutoffs.R): Script used to assess how many ASVs had a taxonomic assignement confidence score below our threshold
[04_precision_and_recall.R](https://github.com/karinevilleneuve/benchmarking_eDNA_method_AlbertaFish/blob/main/04_precision_and_recall.R): Script used to compute the precision and recall score for each method 

# Contact 

For any questions or comments : karine.villeneuve12@gmail.com
